pkg/network/server.go

```diff
@@ -613,12 +621,17 @@ func (s *Server) handleGetRootsCmd(p Peer, gr *payload.GetStateRoots) error {
 
 // handleStateRootsCmd processees `roots` request.
 func (s *Server) handleRootsCmd(rs *payload.StateRoots) error {
-	return nil // TODO
+	for i := range rs.Roots {
+		_ = s.chain.AddStateRoot(&rs.Roots[i])
+	}
+	return nil
 }
 
 // handleStateRootCmd processees `stateroot` request.
 func (s *Server) handleStateRootCmd(r *state.MPTRoot) error {
-	return nil // TODO
+	// we ignore error, because there is nothing wrong if we already have this state root
+	_ = s.chain.AddStateRoot(r)
```

Comment:
It actually should be rebroadcasted if it's something new for us.

Reply:
`StateRoot` is a response for `GetData` command, we have already rebroadcasted it's hash.

Reply:
> StateRoot is a response for GetData command

Not always, see neo-project/neo#1662, but at the same time it's very similar to a block or tx.

> we have already rebroadcasted it's hash.

Where? It should be handled very similar to block (technically even with something similar to `blockQueue` for roots corresponding to not-yet-accepted blocks.

Reply:
Fixed.